src/compton2/Compton2.cc

```diff
@@ -0,0 +1,475 @@
+                      h * bR * bR * (bR - 1)};
+  return H;
+}
+
+Compton2::Compton2(std::string filename)
+    : num_temperatures_(0U), num_groups_(0U), num_leg_moments_(0U), num_evals_(0U), num_points_(0U),
+      Ts_(0U), Egs_(0U), first_groups_(0U), indexes_(0U), data_(0U), derivs_(0U) {
+  Require(filename.length() > 0U);
+  int rank = rtt_c4::node();
+  constexpr int bcast_rank = 0;
+  int errcode = 0;
+
+  if (rank == bcast_rank) {
+    errcode = read_binary(filename);
+  }
+  broadcast_MPI(errcode);
+
+  Ensure(check_class_invariants());
+}
+
+void Compton2::broadcast_MPI(int errcode) {
+  rtt_c4::global_max(errcode);
+  Insist(errcode == 0, "Non-zero errorcode. Exiting.");
+
+  int rank = rtt_c4::node();
+  constexpr int bcast_rank = 0;
+
+  // Broadcast sizes
+  UINT data_size = data_.size();
+  std::array<UINT, 6> pack = {num_temperatures_, num_groups_, num_leg_moments_,
+                              num_evals_,        num_points_, data_size};
+  rtt_c4::broadcast(&pack[0], pack.size(), bcast_rank);
+  UINT p = 0;
+  num_temperatures_ = pack[p++];
+  num_groups_ = pack[p++];
+  num_leg_moments_ = pack[p++];
+  num_evals_ = pack[p++];
+  num_points_ = pack[p++];
+  data_size = pack[p++];
+
+  // Derived sizes
+  const auto tsz = int(num_temperatures_);
+  const auto egsz = int(num_groups_ + 1U);
+  const auto fgsz = int(num_temperatures_ * num_groups_);
+  const auto isz = int(fgsz + 1U);
+  const auto dsz = int(data_size);
+
+  // Broadcast grids
+  if (rank != bcast_rank)
+    Ts_.resize(tsz);
+  rtt_c4::broadcast(&Ts_[0], tsz, 0);
+  if (rank != bcast_rank)
+    Egs_.resize(egsz);
+  rtt_c4::broadcast(&Egs_[0], egsz, 0);
+
+  // Broadcast sparse data structures
+  if (rank != bcast_rank)
+    first_groups_.resize(fgsz);
+  rtt_c4::broadcast(&first_groups_[0], fgsz, 0);
+  if (rank != bcast_rank)
+    indexes_.resize(isz);
+  rtt_c4::broadcast(&indexes_[0], isz, 0);
+
+  // Broadcast data itself
+  if (rank != bcast_rank)
+    data_.resize(dsz);
+  rtt_c4::broadcast(&data_[0], dsz, 0);
+  if (rank != bcast_rank)
+    derivs_.resize(dsz);
+  rtt_c4::broadcast(&derivs_[0], dsz, 0);
+}
+
+int Compton2::read_binary(std::string filename) {
+  // Read
+  auto fin = std::ifstream(filename, std::ios::in | std::ios::binary);
+
+  // Ensure valid type
+  char expected[] = " csk ";
+  char file_type[sizeof(expected)];
+  fin.read(file_type, sizeof(file_type));
+  if (std::strcmp(file_type, expected) != 0) {
+    std::cerr << "Expecting binary file " << filename << " to start with '" << expected
+              << "' but got '" << file_type << "'";
+    std::cerr << std::endl;
+    return 1;
+  }
+
+  UINT binary_ordering, version_major, version_minor;
+  fin.read(reinterpret_cast<char *>(&version_major), sizeof(UINT));
+  fin.read(reinterpret_cast<char *>(&version_minor), sizeof(UINT));
+  fin.read(reinterpret_cast<char *>(&binary_ordering), sizeof(UINT));
+  if (version_major != 1 || binary_ordering > 1) {
```

Comment:
Check for read errors or state of the istream?
It would likely be more efficient to read the whole file into a buffer and then parse the buffer?

Reply:
I defer to you. Should one rank read the file, put it in a buffer, and then MPI communicate that buffer to the other ranks and have everyone parse?